### PR TITLE
Fix leaks with actix_http's client

### PR DESCRIPTION
**Pull request by ptescher**
Each request was creating a new ConnectorPoolSupport future which never
finished. That future also referenced the pool’s inner so even when the
entire pool dropped we leaked the internals. This fixes both.

## PR Type
What kind of change does this PR make?

Bug fix

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.

## Overview
There was a leak in actin-http's client where each request would span a future that would never finish. That future also referenced the pool so the pool would never drop.

This PR fixes both cases. The future now only spawns once when the pool is created and when the pool drops the future finishes.

This should fix:
- https://github.com/actix/actix-web/issues/1523
- https://github.com/actix/actix-web/issues/1464
- https://github.com/actix/actix-web/issues/1428

### CODE REVIEW

**Comment by JohnTitor** [comment]
I don't look in detail yet, but it looks like this breaks some (awc) tests.

**Comment by ptescher** [comment]
@JohnTitor hmm interesting, this shouldn't really affect anything. All the tests pass locally for me.

**Comment by ptescher** [comment]
@JohnTitor ok actually something was out of sync. I see the failures now, will fix.

**Comment by ptescher** [comment]
Ok fixed the tests. One change now is that an actix-rt system needs to be started before a client is created, not just before the first request. This seems fine to me, especially if this is part of a version bump.

**Review by JohnTitor — APPROVED** [review]
I think that makes sense, looks great to me!

**Comment by robjtede** [comment]
think we should get one more eye over this before merging, pretty big deal if this fixes the root cause of those memory leak issues

File: awc/src/request.rs
Comment (cetra3) [inline awc/src/request.rs:589]: I can't see clearly these tests include using multiple clients in parallel.  I am assuming this is where the `ConnectorPool` issues will show up, but I could be wrong.

**Review by cetra3 — APPROVED** [review]
Appears to be a simple change to ensure that the `ConnectorPoolSupport` will eventually resolve the future (if there are no references to it).

I would be interested in seeing a few parallel request tests as well or any test that covers the issues linked to this PR.

**Comment by Dowwie** [comment]
> Ok fixed the tests. One change now is that an actix-rt system needs to be started before a client is created, not just before the first request. This seems fine to me, especially if this is part of a version bump.

@actix/contributors if anyone objects to this change, please share a plausible scenario where there is a conflict

File: awc/src/request.rs
Comment (ptescher) [reply awc/src/request.rs:589]: Because the client now creates a future when it is launched, instead of one per request, these tests now require actix_rt. But the tests themselves don't really use the connectorpool at all.

The benches hit it hard though, since they generate a single client and make multiple requests.

**Comment by ptescher** [comment]
@cetra3 a proper regression test here would check with Tokio to make sure that the number of current futures being worked on is the same before and after a request. There is an open issue for Tokio to support this: https://github.com/tokio-rs/tokio/issues/2510.

**Comment by robjtede** [comment]
> make sure that the number of current futures being worked on is the same before and after a request

That would be a *really* elegant way to detect leaks of this type.

**Comment by cetra3** [comment]
I don't think we should hold up the PR for that change in tokio to land, but maybe another issue raised?

**Comment by robjtede** [comment]
100%. I think get this merged in the next 24hrs if there are no objections from contributors.

**Pull request merged**
